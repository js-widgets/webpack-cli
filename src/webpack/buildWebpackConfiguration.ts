import path from 'path';
import { inspect } from 'util';
import { Configuration } from 'webpack';
import { WidgetDefinition } from 'WidgetDefinition';
import { RegistryConfig } from 'RegistryConfig';
import { SideEffects } from 'common';
import CopyPlugin from 'copy-webpack-plugin';

export default async function buildWebpackConfiguration(
  definitions: WidgetDefinition[],
  configuration: Configuration,
  registryConfig: string,
  outputDir: string,
  logger?: SideEffects,
): Promise<Configuration> {
  let configData: RegistryConfig;
  try {
    const importData = await import(registryConfig);
    configData = importData?.default || importData;
    if (configData.webpackFinal) {
      configuration = await configData.webpackFinal(configuration);
    }
    const { externalPeerDependencies = {} } = configData;
    if (Object.keys(externalPeerDependencies).length) {
      configuration.externals = Object.keys(externalPeerDependencies).reduce(
        (ext, key) => ({
          ...ext,
          [key]: externalPeerDependencies[key].external,
        }),
        {},
      );
      configuration.externalsPresets = { webAsync: true };
      configuration.externalsType = 'window';
    }
  } catch (error: unknown) {
    console.error(error);
  }
  configuration.entry = {};
  for (const definition of definitions) {
    const libName = `render-${definition.shortcode}`.replace(/-./g, (match) =>
      match[1].toUpperCase(),
    );
    configuration.entry[definition.shortcode] = {
      import: definition.entry,
      library: {
        name: libName,
        type: 'window',
        export: 'default',
      },
    };
  }
  if (typeof configuration.output === 'undefined') {
    configuration.output = {};
  }
  configuration.output.path = path.join(outputDir, 'widgets');
  configuration.output.filename =
    configuration.mode === 'production'
      ? '[name]/js/main.[contenthash:8].js'
      : '[name]/js/main.js';
  configuration.output.chunkFilename = '[name]/js/[contenthash:8].chunk.js';
  // Dynamic thumbnail copy.
  const copyOptions = definitions.map((definition) => ({
    from: 'thumbnail.png',
    to: path.join(
      configuration.output?.path || '',
      definition.shortcode,
      '[name].[contenthash:8][ext]',
    ),
    context: path.dirname(definition.entry),
  }));
  configuration.plugins?.push(new CopyPlugin({ patterns: copyOptions }));
  if (logger) {
    logger('\n---------------------------------------------------');
    logger('              Webpack Config                       ');
    logger('---------------------------------------------------');
    logger(inspect(configuration, false, null, true));
    logger('---------------------------------------------------\n\n');
  }
  return configuration;
}
