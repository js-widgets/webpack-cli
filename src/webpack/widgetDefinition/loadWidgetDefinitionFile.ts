import { WidgetDefinition } from 'WidgetDefinition';
import validateWidgetDefinitionFile from './validateWidgetDefinitionFile';
import { dirname, join } from 'path';

/**
 * Loads the widget definition file.
 *
 * @param {string} filename
 */
export default async function loadWidgetDefinitionFile(
  filename: string,
): Promise<WidgetDefinition> {
  const importData = await import(filename);
  // If there is no entry point defined, use `index.tsx` in that folder.
  if (typeof importData.entry === 'undefined') {
    importData.entry = join(dirname(filename), 'index.js');
  }
  validateWidgetDefinitionFile(importData);
  return importData;
}
