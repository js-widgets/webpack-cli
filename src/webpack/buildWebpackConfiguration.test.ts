import buildWebpackConfiguration from './buildWebpackConfiguration';
import loadWidgetDefinitionFile from './widgetDefinition/loadWidgetDefinitionFile';
import { join as pathJoin } from 'path';

describe('buildWebpackConfiguration', () => {
  it('builds the correct configuration with empty data', async () => {
    expect.assertions(1);
    const actual = await buildWebpackConfiguration([], {}, '', '');
    expect(actual).toMatchSnapshot();
  });
  it('builds the correct configuration', async () => {
    expect.assertions(1);
    const definitions = [
      await loadWidgetDefinitionFile(
        pathJoin(__dirname, '../__testData__/widgetDefinitions/fake.widget'),
      ),
    ];
    const actual = await buildWebpackConfiguration(
      definitions,
      {},
      pathJoin(
        __dirname,
        '../__testData__/validProject/.widgetRegistry/main.js',
      ),
      './src/__testData__/widgetDefinitions/',
    );
    expect(actual).toMatchSnapshot();
  });
});
