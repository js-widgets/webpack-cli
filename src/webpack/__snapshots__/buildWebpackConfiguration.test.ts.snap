// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildWebpackConfiguration builds the correct configuration 1`] = `
Object {
  "devtool": "inline-source-map",
  "entry": Object {
    "fake": Object {
      "import": "./render.js",
      "library": Object {
        "export": "default",
        "name": "renderFake",
        "type": "window",
      },
    },
  },
  "externals": Object {
    "react": "React",
    "react-dom": "ReactDOM",
    "react-intl": "ReactIntl",
    "react-slick": "Slider",
  },
  "externalsPresets": Object {
    "webAsync": true,
  },
  "externalsType": "window",
  "mode": "none",
  "module": Object {
    "rules": Array [
      Object {
        "test": /\\\\\\.\\(s\\[ac\\]\\|c\\)ss\\$/i,
        "use": Array [
          "style-loader",
          "css-loader",
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "postcssOptions": Object {
                "plugins": Array [
                  Array [
                    "postcss-preset-env",
                    Object {},
                  ],
                ],
              },
            },
          },
          "sass-loader",
        ],
      },
      Object {
        "test": /\\\\\\.\\(png\\|jpe\\?g\\|gif\\)\\$/i,
        "type": "asset",
      },
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.\\(t\\|j\\)sx\\?\\$/,
        "use": Object {
          "loader": "babel-loader",
          "options": Object {
            "cacheDirectory": true,
            "presets": Array [
              "@babel/preset-env",
              "@babel/preset-typescript",
            ],
          },
        },
      },
    ],
  },
  "output": Object {
    "assetModuleFilename": "[name]/images/[hash][ext][query]",
    "chunkFilename": "[name]/js/[contenthash:8].chunk.js",
    "filename": "[name]/js/main.js",
    "path": "src/__testData__/widgetDefinitions/widgets",
  },
  "plugins": Array [
    CleanWebpackPlugin {
      "apply": [Function],
      "cleanAfterEveryBuildPatterns": Array [],
      "cleanOnceBeforeBuildPatterns": Array [
        "**/*",
      ],
      "cleanStaleWebpackAssets": true,
      "currentAssets": Array [],
      "dangerouslyAllowCleanPatternsOutsideProject": false,
      "dry": false,
      "handleDone": [Function],
      "handleInitial": [Function],
      "initialClean": false,
      "outputPath": "",
      "protectWebpackAssets": true,
      "removeFiles": [Function],
      "verbose": false,
    },
    EnvironmentPlugin {
      "defaultValues": Object {
        "DEBUG": false,
        "NODE_ENV": "development",
        "PUBLIC_ASSETS_URL": "",
        "PUBLIC_URL": "",
      },
      "keys": Array [
        "NODE_ENV",
        "DEBUG",
        "PUBLIC_URL",
        "PUBLIC_ASSETS_URL",
      ],
    },
    CopyPlugin {
      "options": Object {},
      "patterns": Array [
        Object {
          "context": ".",
          "from": "thumbnail.png",
          "to": "src/__testData__/widgetDefinitions/widgets/fake/[name].[contenthash:8][ext]",
        },
      ],
    },
  ],
  "profile": true,
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".jsx",
      ".ts",
      ".tsx",
    ],
  },
  "target": "web",
}
`;

exports[`buildWebpackConfiguration builds the correct configuration with empty data 1`] = `
Object {
  "devtool": "inline-source-map",
  "entry": Object {},
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "test": /\\\\\\.\\(s\\[ac\\]\\|c\\)ss\\$/i,
        "use": Array [
          "style-loader",
          "css-loader",
          Object {
            "loader": "postcss-loader",
            "options": Object {
              "postcssOptions": Object {
                "plugins": Array [
                  Array [
                    "postcss-preset-env",
                    Object {},
                  ],
                ],
              },
            },
          },
          "sass-loader",
        ],
      },
      Object {
        "test": /\\\\\\.\\(png\\|jpe\\?g\\|gif\\)\\$/i,
        "type": "asset",
      },
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.\\(t\\|j\\)sx\\?\\$/,
        "use": Object {
          "loader": "babel-loader",
          "options": Object {
            "cacheDirectory": true,
            "presets": Array [
              "@babel/preset-env",
              "@babel/preset-typescript",
            ],
          },
        },
      },
    ],
  },
  "output": Object {
    "assetModuleFilename": "[name]/images/[hash][ext][query]",
    "chunkFilename": "[name]/js/[contenthash:8].chunk.js",
    "filename": "[name]/js/main.js",
    "path": "widgets",
  },
  "plugins": Array [
    CleanWebpackPlugin {
      "apply": [Function],
      "cleanAfterEveryBuildPatterns": Array [],
      "cleanOnceBeforeBuildPatterns": Array [
        "**/*",
      ],
      "cleanStaleWebpackAssets": true,
      "currentAssets": Array [],
      "dangerouslyAllowCleanPatternsOutsideProject": false,
      "dry": false,
      "handleDone": [Function],
      "handleInitial": [Function],
      "initialClean": false,
      "outputPath": "",
      "protectWebpackAssets": true,
      "removeFiles": [Function],
      "verbose": false,
    },
    EnvironmentPlugin {
      "defaultValues": Object {
        "DEBUG": false,
        "NODE_ENV": "development",
        "PUBLIC_ASSETS_URL": "",
        "PUBLIC_URL": "",
      },
      "keys": Array [
        "NODE_ENV",
        "DEBUG",
        "PUBLIC_URL",
        "PUBLIC_ASSETS_URL",
      ],
    },
  ],
  "profile": true,
  "resolve": Object {
    "extensions": Array [
      ".js",
      ".jsx",
      ".ts",
      ".tsx",
    ],
  },
  "target": "web",
}
`;
